### DISCUSSION TITLE
MICR EXPORTCLONE v4.1 unknow --preset option

### DISCUSSION THREAD

**Discussion by GregoireDS** [milaboratory/mixcr#883] [Q&A]
Hi,

We've updated to the latest version of MIXCR. Since then, I'm struggling with an error with **_mixcr exportclone_**.

Error concerns --preset

In older version I used : **_mixcr exportClones -p fullImputed_**

However, In the new version I get the following error :

**Unknown options: '-p',
Possible solutions: --prepend-columns, -positionInReferenceOf, -positionOf**

Could you indicate me whether _**"--preset"**_ option has changed within the new version?

Thanks

-----------

Processing  elidEt_168
mixcr.jar align
WARNING: unnecessary override -OallowPartialAlignments=true with the same value.
The following analysis parameters will be used:
species: "hsa"
libraryName: "default"
trimmingQualityThreshold: 0
trimmingWindowSize: 6
chains: "ALL"
overlapPairedReads: true
bamDropNonVDJ: false
writeFailedAlignments: false
tagPattern: null
tagUnstranded: false
tagMaxBudget: 10.0
limit: null
parameters:
  vParameters:
    geneFeatureToAlign: "VTranscriptWithP"
    minSumScore: 80
    relativeMinScore: 0.97
    parameters:
      type: "kaligner"
      mapperKValue: 5
      floatingLeftBound: false
      floatingRightBound: true
      mapperAbsoluteMinScore: 1.5
      mapperRelativeMinScore: 0.7
      mapperMatchScore: 1.0
      mapperMismatchPenalty: -0.1
      mapperOffsetShiftPenalty: -0.3
      mapperMinSeedsDistance: 4
      mapperMaxSeedsDistance: 10
      minAlignmentLength: 15
      maxAdjacentIndels: 2
      alignmentStopPenalty: -1000
      absoluteMinScore: 55.0
      relativeMinScore: 0.87
      maxHits: 7
      scoring:
        type: "linear"
        alphabet: "nucleotide"
        subsMatrix: "simple(match = 5, mismatch = -11)"
        gapPenalty: -21
  dParameters:
    geneFeatureToAlign: "DRegionWithP"
    relativeMinScore: 0.85
    absoluteMinScore: 25.0
    maxHits: 3
    scoring:
      type: "linear"
      alphabet: "nucleotide"
      subsMatrix: "simple(match = 5, mismatch = -11)"
      gapPenalty: -21
  jParameters:
    geneFeatureToAlign: "JRegionWithP"
    minSumScore: 60
    relativeMinScore: 0.97
    parameters:
      type: "kaligner"
      mapperKValue: 5
      floatingLeftBound: true
      floatingRightBound: false
      mapperAbsoluteMinScore: 1.5
      mapperRelativeMinScore: 0.75
      mapperMatchScore: 1.0
      mapperMismatchPenalty: -0.1
      mapperOffsetShiftPenalty: -0.3
      mapperMinSeedsDistance: 4
      mapperMaxSeedsDistance: 8
      minAlignmentLength: 15
      maxAdjacentIndels: 2
      alignmentStopPenalty: -1000
      absoluteMinScore: 50.0
      relativeMinScore: 0.87
      maxHits: 7
      scoring:
        type: "linear"
        alphabet: "nucleotide"
        subsMatrix: "simple(match = 5, mismatch = -11)"
        gapPenalty: -21
  cParameters:
    geneFeatureToAlign: "CExon1"
    minSumScore: 40
    relativeMinScore: 0.97
    parameters:
      type: "kaligner"
      mapperKValue: 5
      floatingLeftBound: false
      floatingRightBound: false
      mapperAbsoluteMinScore: 1.5
      mapperRelativeMinScore: 0.75
      mapperMatchScore: 1.0
      mapperMismatchPenalty: -0.1
      mapperOffsetShiftPenalty: -0.3
      mapperMinSeedsDistance: 4
      mapperMaxSeedsDistance: 10
      minAlignmentLength: 15
      maxAdjacentIndels: 2
      alignmentStopPenalty: -1000
      absoluteMinScore: 40.0
      relativeMinScore: 0.87
      maxHits: 7
      scoring:
        type: "linear"
        alphabet: "nucleotide"
        subsMatrix: "simple(match = 5, mismatch = -11)"
        gapPenalty: -21
  vjAlignmentOrder: "VThenJ"
  libraryStructure: "Unknown"
  includeDScore: true
  includeCScore: true
  minSumScore: 200.0
  maxHits: 5
  relativeMinVFR3CDR3Score: 0.7
  allowPartialAlignments: true
  allowNoCDR3PartAlignments: false
  allowChimeras: false
  readsLayout: "Opposite"
  mergerParameters:
    qualityMergingAlgorithm: "MaxSubtraction"
    partsLayout: null
    minimalOverlap: 17
    maxQuality: 50
    minimalIdentity: 0.9
    identityType: "Unweighted"
  fixSeed: true
  alignmentBoundaryTolerance: 5
  minChimeraDetectionScore: 120
  vjOverlapWindow: 3
  saveOriginalSequence: false
  saveOriginalReads: false
  smartForceEdgeAlignments: true
====================
WARNING: 63 non-functional genes excluded.
Alignment: 0%
Alignment: 10.7%  ETA: 00:01:15
Alignment: 21.3%  ETA: 00:00:59
Alignment: 32%  ETA: 00:00:57
Alignment: 43.7%  ETA: 00:00:44
Alignment: 54.4%  ETA: 00:00:21
Alignment: 65%  ETA: 00:00:16
Alignment: 75.7%  ETA: 00:00:11
Alignment: 87.4%  ETA: 00:00:06
Alignment: 98.1%  ETA: 00:00:00
============== Align Report ==============
Analysis time: 1.06m
Total sequencing reads: 1212437
Successfully aligned reads: 771 (0.06%)
Paired-end alignment conflicts eliminated: 4 (0%)
Alignment failed, no hits (not TCR/IG?): 1182945 (97.57%)
Alignment failed because of absence of CDR3 parts: 8797 (0.73%)
Alignment failed because of low total score: 19924 (1.64%)
Overlapped: 822561 (67.84%)
Overlapped and aligned: 505 (0.04%)
Alignment-aided overlaps: 37 (7.33%)
Overlapped and not aligned: 822056 (67.8%)
No CDR3 parts alignments, percent of successfully aligned: 19 (2.46%)
Partial aligned reads, percent of successfully aligned: 318 (41.25%)
TRA chains: 154 (19.97%)
TRA non-functional: 0 (0%)
TRB chains: 555 (71.98%)
TRB non-functional: 0 (0%)
TRAD chains: 62 (8.04%)
TRAD non-functional: 0 (0%)
Realigned with forced non-floating bound: 779826 (64.32%)
Realigned with forced non-floating right bound in left read: 280 (0.02%)
Realigned with forced non-floating left bound in right read: 280 (0.02%)
mixcr.jar assemblePartial
The following analysis parameters will be used:
overlappedOnly: false
dropPartial: false
cellLevel: false
parameters:
  kValue: 12
  kOffset: -7
  minimalAssembleOverlap: 12
  minimalNOverlap: 7
  minimalAlignmentMergeIdentity: 0.85
  mergerParameters:
    qualityMergingAlgorithm: "SumSubtraction"
    partsLayout: "CollinearDirect"
    minimalOverlap: 20
    maxQuality: 45
    minimalIdentity: 0.95
    identityType: "Unweighted"
  maxLeftParts: 256000
  maxLeftMatches: 6144
====================
Building index: 0%
Searching for overlaps: 0%
============== AssemblePartial Report ==============
Analysis time: 1.92s
Total alignments analysed: 771
Number of output alignments: 737 (95.59%)
Alignments already with CDR3 (no overlapping is performed): 431 (55.9%)
Successfully overlapped alignments: 34 (4.41%)
Left parts with too small N-region (failed to extract k-mer): 82 (10.64%)
Extracted k-mer diversity: 601
Dropped due to wildcard in k-mer: 0 (0%)
Dropped due to too short NRegion parts in overlap: 47 (6.1%)
Dropped overlaps with empty N region due to no complete NDN coverage: 0 (0%)
Number of left-side alignments: 111 (14.4%)
Number of right-side alignments: 163 (21.14%)
Complex overlaps: 0 (0%)
Over-overlaps: 0 (0%)
Partial alignments written to output: 272 (35.28%)
mixcr.jar assemble
The following analysis parameters will be used:
sortBySequence: false
clnaOutput: true
cellLevel: false
consensusAssemblerParameters: null
cloneAssemblerParameters:
  assemblingFeatures:
    - "CDR3"
  minimalClonalSequenceLength: 12
  qualityAggregationType: "Max"
  cloneClusteringParameters:
    searchDepth: 2
    allowedMutationsInNRegions: 1
    minimalTagSetOverlap: -1.0
    searchParameters: "twoMismatchesOrIndels"
    clusteringFilter:
      type: "relativeConcentration"
      specificMutationProbability: 0.001
  cloneFactoryParameters:
    vParameters:
      relativeMinScore: null
      scoring: null
      maxAlignmentWidthLinear: 5
      maxAlignmentWidthAffine: 500
    jParameters:
      relativeMinScore: null
      scoring: null
      maxAlignmentWidthLinear: 5
      maxAlignmentWidthAffine: 500
    cParameters:
      relativeMinScore: null
      scoring: null
      maxAlignmentWidthLinear: 5
      maxAlignmentWidthAffine: 500
    dParameters:
      relativeMinScore: null
      absoluteMinScore: null
      maxHits: null
      scoring: null
    vdcParameters:
      Variable:
        relativeMinScore: null
        scoring: null
        maxAlignmentWidthLinear: 5
        maxAlignmentWidthAffine: 500
      Joining:
        relativeMinScore: null
        scoring: null
        maxAlignmentWidthLinear: 5
        maxAlignmentWidthAffine: 500
      Constant:
        relativeMinScore: null
        scoring: null
        maxAlignmentWidthLinear: 5
        maxAlignmentWidthAffine: 500
  separateByV: false
  separateByJ: false
  separateByC: false
  maximalPreClusteringRatio: 1.0
  preClusteringScoreFilteringRatio: 2.0
  preClusteringCountFilteringRatio: 2.0
  addReadsCountOnClustering: false
  badQualityThreshold: 20
  maxBadPointsPercent: 0.7
  mappingThreshold: "2of5"
  minimalQuality: 0
  postFilters: null
inferMinRecordsPerConsensus: false
====================
Initialization: progress unknown
Initialization: progress unknown
============== Assemble Report ==============
Analysis time: 1.76s
Final clonotype count: 2
Average number of reads per clonotype: 244
Reads used in clonotypes, percent of total: 488 (0.04%)
Reads used in clonotypes before clustering, percent of total: 491 (0.04%)
Number of reads used as a core, percent of used: 479 (97.56%)
Mapped low quality reads, percent of used: 12 (2.44%)
Reads clustered in PCR error correction, percent of used: 3 (0.61%)
Reads pre-clustered due to the similar VJC-lists, percent of used: 0 (0%)
Reads dropped due to the lack of a clone sequence, percent of total: 272 (0.02%)
Reads dropped due to a too short clonal sequence, percent of total: 0 (0%)
Reads dropped due to low quality, percent of total: 0 (0%)
Reads dropped due to failed mapping, percent of total: 8 (0%)
Reads dropped with low quality clones, percent of total: 0 (0%)
Clonotypes eliminated by PCR error correction: 1
Clonotypes dropped as low quality: 0
Clonotypes pre-clustered due to the similar VJC-lists: 0
Clones dropped in post filtering: 0 (0%)
Reads dropped in post filtering: 0.0 (0%)
Alignments filtered by tag prefix: 0 (0%)
TRA chains: 1 (50%)
TRA non-functional: 0 (0%)
TRB chains: 1 (50%)
TRB non-functional: 0 (0%)
mixcr.jar assembleContigs
The following analysis parameters will be used:
ignoreTags: false
parameters:
  branchingMinimalQualityShare: 0.1
  branchingMinimalSumQuality: 60
  decisiveBranchingSumQualityThreshold: 120
  alignedSequenceEdgeDelta: 3
  alignmentEdgeRegionSize: 7
  minimalNonEdgePointsFraction: 0.25
  minimalMeanNormalizedQuality: 5.0
  outputMinimalQualityShare: 0.75
  outputMinimalSumQuality: 0
  subCloningRegions: null
  assemblingRegions: null
  postFiltering:
    type: "NoFiltering"
  trimmingParameters:
    averageQualityThreshold: 10.0
    windowSize: 8
  minimalContigLength: 20
  alignedRegionsOnly: false
====================
Assembling contigs: 0%
============== AssembleContigs Report ==============
Analysis time: 2.05s
Initial clonotype count: 2
Final clonotype count: 2 (100%)
Canceled assemblies: 0 (0%)
Number of premature termination assembly events, percent of number of initial clonotypes: 0.0 (0%)
Longest contig length: 799
Clustered variants: 0 (0%)
Reads in clustered variants: 0.0 (0%)
Reads in divided (newly created) clones: 0.0 (0%)
Clones with ambiguous letters in splitting region: 0 (0%)
Reads in clones with ambiguous letters in splitting region: 0.0 (0%)
Average number of ambiguous letters per clone with ambiguous letters in splitting region: NaN
mixcr.jar exportClones CLNSTRA
Unknown options: '-p', '/data/gece/gdestreel/scRNAseq_SMARTseq2/Run_for_ORB_and_ALIX_October_2022/Outputs/ELIDET_HN00142131/Mixcr3/Raw_Results/elidEt_168_clones_TRA.txt'
Possible solutions: --prepend-columns, -positionInReferenceOf, -positionOf

**Comment by PoslavskySV** [comment]
Hi, 

the old 'preset' option used in export is now superseded by the global [analysis preset](https://docs.milaboratories.com/mixcr/reference/overview-presets/), which also defines the export fields as well. If you want to modify your export, you can do this with mixins or write your own export preset and use it. In case if you just looking for imputation you can use the following mixin option:

```
mixcr exportClones --impute-germline-on-export file.clns file.tsv
```

Let me better know what is your case and we'll provide additional instructions.

**Reply by GregoireDS** [reply]
Thanks for the reply,

Ok I see know that the hierarchy has changed a bit. If I understood it correctly, presets should be defined in _**mixcr analyze**_ only. However this is specific to the case I want to run the whole pipeline through **_mixcr analyze**.
In my case,  I'd rather like to run **_mixcr align_** **_mixcr assemble_** and _**mixcr exportClones**_ separetely.  In this scenario, when I run the modified (according to your comment) command : 

**_java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar exportClones --impute-germline-on-export --filter-out-of-frames --filter-stops -c TRB ${CLNS} ${CLNSTRB}_**

I get the following error : 

**_Unknown option: '--impute-germline-on-export'_** followed by the whole documentation on mixcr export Clones (Usage: java -jar mixcr.jar exportClones). 

Does it means that the **_--impute-germline-on-export_**  option is not available ? 

Thanks for your help

**Comment by mizraelson** [comment]
Hi,
Actually you can specify the presets on `mixcr align` step (the preset may include settings for all other functions as well including `exportClones`). So the idea is that you run `mixcr align` with a provided preset, and then when you run downstream commands they will automatically use the original preset.

So in your case there are two options:
1) You can create a custom preset for all steps so you wouldn't have to provide any settings later and provide it to `mixcr align`. If you can share the exact commands with all the options you use, i can help you with creating the preset. You could even use this preset with analyze command if you want and it will run the whole pipeline specifically tuned for your needs.
2) You can use `-nFeatureImputed VDJRegion -aaFeatureImputed VDJRegion` parameters, where `VDJRegion` is the feature you want to impute. You can pick any set of features.
example:

`mixcr exportClones -nFeatureImputed VDJRegion -aaFeatureImputed VDJRegion --filter-out-of-frames --filter-stops input.clns output.tsv`

`--impute-germline-on-export` option is in develop and will be released soon.

**Comment by GregoireDS** [comment]
Hi,

Thanks for your very clear answer ! 

These are my commands with options, could you help me creating presets for **mixcr align** ?

Have a nice day,

**_echo "mixcr.jar align"
java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar align -t 72 --preset rnaseq-tcr-full-length -s hsa -OallowPartialAlignments=true -r ${PROJECTDIR}/${FOLDER}/Mixcr3/Raw_Results/${SAMPLE}_mixcr_align_report.txt  ${I1} ${I2} ${VDJCA} --verbose 

echo "mixcr.jar assemblePartial"
java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar assemblePartial ${VDJCA} ${RVDJCA}

echo "mixcr.jar assemble"
java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar assemble --write-alignments ${RVDJCA} ${CLNA}

echo "mixcr.jar assembleContigs"
java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar assembleContigs ${CLNA} ${CLNS}

echo "mixcr.jar exportClones CLNSTRA"
java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar exportClones --filter-out-of-frames --filter-stops -c TRA ${CLNS} ${CLNSTRA}

echo "mixcr.jar exportClones CLNSTRB"
java -Xms64g -Xmx64g -jar ${MIXCRPATH}/mixcr.jar exportClones --filter-out-of-frames --filter-stops -c TRB ${CLNS} ${CLNSTRB}_**
